cosmos/web/views.py

```diff
@@ -57,23 +58,126 @@ def workflow(name):
         workflow = session.query(Workflow).filter_by(name=name).one()
         return render_template("cosmos/workflow.html", workflow=workflow)
 
-    @bprint.route("/workflow/<workflow_name>/<stage_name>/")
+    @bprint.route("/workflow/<workflow_name>/<stage_name>/", methods=['GET', 'POST'])
     def stage(workflow_name, stage_name):
+        IN_PAGE = 40  # how many tasks to show on one page
+        # these are the column names that appear in the screen
+        colnames = ["id", "task", "successful", "status", "drm_status", "drm_jobid",
+                    "attempts", "submitted_on", "finished_on", "wall_time"]
+        # this indicates if this column can be used for sorting and searching
+        names_good = ["id", "params", "successful", False, False, False,
+                      "attempt", "submitted_on", "finished_on", "wall_time"]
+
+        page = request.args.get('page', 1, type=int)
+        keyword = request.args.get('keyword', '', type=str)
+        sorting = request.args.get('sorting', None, type=str)
+        order = request.args.get('order', None, type=str)
+
         ex = session.query(Workflow).filter_by(name=workflow_name).one()
         stage = session.query(Stage).filter_by(workflow_id=ex.id, name=stage_name).one()
         if stage is None:
             return abort(404)
-        submitted = [t for t in stage.tasks if t.status == TaskStatus.submitted]
+        from sqlalchemy import text
+        tasks = session.query(Task).filter_by(stage_id=stage.id)
+        # search keyword
+        if keyword == '':
+            tasks_searched = tasks
+        else:
+            pattern = keyword.replace("'", "''")
+            tasks_searched = tasks.filter(or_(
+                *[getattr(Task, field).like(f"%{keyword}%") if field != 'params' else None
+                  for field in Task.__searchable__], text(f"params LIKE '%{pattern}%'")
+                ))
+
+        # sort
+        tasks_sorted = tasks_searched
+        if sorting is not None:
```

Comment:
i should have been clearer, but just having one sort (either ascending or descending of a single column) would be sufficient

Reply:
so should I remove the descending order?

Reply:
descending is still nice, thanks